Adding ShowHideToggle component

georgewrmarshall opened the pull request:
## Explanation

Adding ShowHideToggle component and eye icon

<img width="1440" alt="Screen Shot 2022-03-15 at 8 17 14 AM" src="https://user-images.githubusercontent.com/8112138/158411441-80ed420f-114a-4262-ad6a-d677f6ebe7e3.png">
<img width="1434" alt="Screen Shot 2022-03-15 at 8 21 20 AM" src="https://user-images.githubusercontent.com/8112138/158411446-709c20d3-64b7-4f06-8d13-bf279410a3de.png">
<img width="1435" alt="Screen Shot 2022-03-15 at 8 21 36 AM" src="https://user-images.githubusercontent.com/8112138/158411450-dfbdb1c6-e266-48a7-98ce-9dd5dbccc4b0.png">

Gudahtt committed:
Initial implementation of new SrpInput component

This new version of the SrpInput component uses a separate field for
each word of the SRP. Only one field can be revealed at a time, making
it less likely that it gets accidentally revealed to somebody.

Gudahtt committed:
Fix copy mistakes

Gudahtt committed:
Move container div from 'create vault' to 'srp-input', and setup grid layout

Gudahtt committed:
Increase size of title

Gudahtt committed:
Remove hard-coded width in Storybook to allow testing different viewport sizes

Gudahtt committed:
Improve layout

Gudahtt committed:
Improve margins

Gudahtt committed:
Update dropdown text

Gudahtt committed:
Expand SRP input section

georgewrmarshall committed:
Adding ShowHideToggle component

georgewrmarshall committed:
Updates to component api in srp and color of icon

Gudahtt commented on ui/components/ui/icon/icon-eye.js:
```diff
@@ -0,0 +1,42 @@
+    xmlns="http://www.w3.org/2000/svg"
+    viewBox="0 0 512 512"
+  >
+    <path d="m444 247c-36-71-108-119-190-119-82 0-154 48-190 119-1 2-2 6-2 9 0 4 1 7 2 10 36 71 108 118 190 118 82 0 154-47 190-118 1-3 2-6 2-9 0-4-1-8-2-10z m-190 105c-53 0-96-43-96-96 0-53 43-96 96-96 53 0 96 43 96 96l0 1c0 52-43 96-96 96z m0-160c-6 1-12 1-17 3 4 5 6 12 6 19 0 17-14 31-31 31-8 0-14-2-20-6-1 6-2 12-2 17 0 35 29 64 64 64 36 0 64-29 64-64 0-35-28-63-64-63z" />
```
Where did this SVG come from? Is this fontawesome?

georgewrmarshall replied:
Yes, I just grabbed it from the Figma file but I wasn't sure where it was from. I think ideally we are using svgs over font icons. But do you think this could be breaking a usage agreement?

Gudahtt replied:
If it's from the "FontAwesome Free" pack, this is allowed under their creative commons license. We're allowed to embed it and change it however we like. But it's dangerous to embed IP we don't own directly in our product without documenting what it is or where it came from. _Minimally_ we should document what it is and where it came from. That way we can be assured that it's allowed.

Gudahtt replied:
I'd strongly recommend against _ever_ copying an SVG directly from Figma. Our Figma designs have included "FontAwesome Pro" icons in the past (which we cannot use, because our usage is incompatible with their license, they won't even let us pay for it), and they have contained other proprietary icons from other sources.  Big no-no. Please don't ever do this ,and speak up if you ever see somebody else doing it.

georgewrmarshall replied:
OK thanks. That's good to know. This should be flagged at the design level or at least before design handover to dev. Is this common knowledge in design? Might be worth reminding them. As for using font icons they can render unexpectedly in different browsers so we should be pushing for svg icons. I've seen a few features using svgs go into the codebase with svgs being pulled from Figma so I think we should definitely flag this in the right channels.

Gudahtt replied:
We do have all of the SVG icons as a dependency (`@fortawesome/fontawesome-free`, see the `svgs` directory). These are the official SVGs. We could copy them direct from there and leave a comment saying where they came from (including the version). Or we could use the FontAwesome React library, where they've already packaged each SVG as a component. See here for details: https://fontawesome.com/v6/docs/web/use-with/react/

Gudahtt replied:
I copied over the official SVGs and added a comment explaining where we got them

Gudahtt commented on ui/components/ui/show-hide-toggle/show-hide-toggle.js:
```diff
@@ -0,0 +1,65 @@
+        disabled={disabled}
+      />
+      <label htmlFor={id} className="show-hide-toggle__label">
+        <IconEye ariaLabel={ariaLabel} className="show-hide-toggle__icon" />
```
Looks like this doesn't display any visual indication when it is focused. Or maybe it's the checkbox that is getting focused, and it's not shown because it's off screen. Either way, we have to ensure there is a visual representation of focus for the sake of helping keyboard users.

Curious why you chose to go this route of separating the checkbox component from the label, hiding the checkbox, and using the label instead. This seems like a lot more work than using CSS to display the checkbox as an eye. It's not clear this is any better for accessibility, readability, or even compatibility.

georgewrmarshall replied:
I should double check this is an acceptable standard for a custom checkbox but I believed it was. I do see a visual representation of focus in my storybook do you see something different. I did change the component api in another update


https://user-images.githubusercontent.com/8112138/158444898-ba3d5072-4b14-4f85-b713-bf11341ac4b8.mov

georgewrmarshall replied:
No haven't browser tested and haven't written unit tests yet. I'll do that now

Gudahtt replied:
This was fixed by replacing the webkit-specific highlight color referenced in the styles with `Highlight`, which should work across browsers.

Also George explained to me the evils of icon fonts, so I'm now convinced that this is a much better approach than what I had put together before this :sweat_smile:

Gudahtt committed:
Remove unused localized messages

Gudahtt committed:
Update dropdown option names in unit tests

Gudahtt committed:
Add official SVG icons for `eye` and `eye-slash` and fix focus

The icons from the official FontAwesome-free package were copied into
the `EyeIcon` component. Adding the `eye-slash` variant allows
simplifying the `ShowHideToggle` component, as any CSS for the custom
slask could now be removed.

Focus is fixed on Firefox by using the color `Highlight` rather than
the webkit-specific color used previously. `Highlight` apparently is
standardized (it was briefly deprecated but that was undone, it is no
longer deprecated today), and is the color the browser uses to show
something is selected.

Some CSS rules have also been removed or simplified. The checkbox
element is now hidden with `appearance: none` instead of moving it off-
screen.

georgewrmarshall commented on ui/components/ui/icon/icon-eye.js:
```diff
@@ -6,17 +6,24 @@ const IconEye = ({
     aria-label={ariaLabel}
     xmlns="http://www.w3.org/2000/svg"
-    viewBox="0 0 512 512"
+    viewBox="0 0 640 512"
```
What is the purpose of changing the viewBox here?

Gudahtt replied:
The fontawesome SVGs each used a different viewbox. Both were an increase from the Figma version seemingly, but they had different widths (the eye-slash one was slightly wider), so I went with the wider one to ensure it didn't get cut off.

As we just discussed I'm going to move this to a separate component, so I'll restore the exact viewbox that the SVGs use.

georgewrmarshall commented on ui/components/ui/icon/icon-eye.js:
```diff
@@ -6,17 +6,24 @@ const IconEye = ({
+    ) : (
+      <path d="M288 144a110.94 110.94 0 0 0-31.24 5 55.4 55.4 0 0 1 7.24 27 56 56 0 0 1-56 56 55.4 55.4 0 0 1-27-7.24A111.71 111.71 0 1 0 288 144zm284.52 97.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400c-98.65 0-189.09-55-237.93-144C98.91 167 189.34 112 288 112s189.0 ...
+    )}
   </svg>
```
The intention was that would simply be the eye icon as it could be used in different context.

Gudahtt replied:
Resolved by splitting into two components.

georgewrmarshall reviewed (commented):
Left some comments.

Gudahtt committed:
Split eye icon into `IconEye` and `IconEyeSlash`

The slash and unslashed eye have been split into separate components.

Gudahtt committed:
Add/update various props

The following changes have been made:

* Rename `dataTestId` prop to `data-testid`
  See the discussion in `#metamask-dev` about this. The thinking was
that `data-testid` is what this is called for any native React
component, or for any "wrapper" component that passes through all extra
props. So we may as well follow that same convention to make our
components easier to use.
* Rename "checked" to "shown"
  This prop represents whether the Show/Hide toggle is in the "shown'
state or not, so this seemed more appropriate. The checkbox state is
kindof an implementation detail.
* Split `areaLabel` prop into `ariaLabelHidden` and `ariaLabelShown`
  The `aria-label` attribute should describe what the icon represents,
but the icon we use in the show and hide case has a different meaning
(show and hide respectively). We now accept one label for each case.
This makes testing easier as well.
* Add `title` attribute
  This lets us describe the purpose of the toggle. When applied to the
`<label>`, it shows on hover.

Gudahtt committed:
Add unit tests

Gudahtt committed:
Update SrpInput component to reflect recent changes

The SrpInput component has been updated to account for recent changes
to the props of the `ShowHideToggle` component. New required localized
messages have been added as well.

georgewrmarshall commented:
Closing as this is addressed in https://github.com/MetaMask/metamask-extension/pull/13979
